typescript/index.d.ts

```diff
@@ -0,0 +1,264 @@
+ */
+export function useRecoilValueLoadable<T>(recoilValue: RecoilValue<T>): Loadable<T>;
+
+/**
+ * Returns a tuple where the first element is the value of the recoil state
+ * and the second is a setter to update that state. Subscribes component
+ * to updates of the given state.
+ */
+export function useRecoilState<T>(recoilState: RecoilState<T>): [T, SetterOrUpdater<T>];
+
+/**
+ * Returns a tuple where the first element is a Loadable and the second
+ * element is a setter function to update the given state. Subscribes
+ * component to updates of the given state.
+ */
+export function useRecoilStateLoadable<T>(recoilState: RecoilState<T>): [Loadable<T>, SetterOrUpdater<T>];
+
+/**
+ * Returns a setter function for updating Recoil state. Does not subscribe
+ * the component to the given state.
+ */
+
+export function useSetRecoilState<T>(recoilState: RecoilState<T>): SetterOrUpdater<T>;
+
+/**
+ * Returns a function that will reset the given state to its default value.
+ */
+export function useResetRecoilState(recoilState: RecoilState<any>): Resetter;
+
+/**
+ * Returns a function that will run the callback that was passed when
+ * calling this hook. Useful for accessing Recoil state in response to
+ * events.
+ */
+export function useRecoilCallback<Args extends ReadonlyArray<unknown>, Return>(
+  fn: (interface: CallbackInterface, ...args: Args) => Return,
+  deps?: ReadonlyArray<unknown>,
+): (...args: Args) => Return;
+
+// loadable.d.ts
+export type ResolvedLoadablePromiseInfo<T> = Readonly<{
+  value: T;
+  upstreamState__INTERNAL_DO_NOT_USE?: TreeState;
+}>;
+
+export type LoadablePromise<T> = Promise<ResolvedLoadablePromiseInfo<T>>;
+
+export type Loadable<T> =
+  | Readonly<{
+      state: 'hasValue';
+      contents: T;
+    }>
+  | Readonly<{
+      state: 'hasError';
+      contents: Error;
+    }>
+  | Readonly<{
+      state: 'loading';
+      contents: LoadablePromise<T>;
+    }>;
+
+// recoilValue.d.ts
+export class AbstractRecoilValue<T> {
+  __tag: [T];
+  __cTag: (t: T) => void; // for contravariance
+
+  key: NodeKey;
+  constructor(newKey: NodeKey);
+}
+
+export class AbstractRecoilValueReadonly<T> {
+  __tag: [T];
+
+  key: NodeKey;
+  constructor(newKey: NodeKey);
+}
+
+export class RecoilState<T> extends AbstractRecoilValue<T> {}
+export class RecoilValueReadOnly<T> extends AbstractRecoilValueReadonly<T> {}
+export type RecoilValue<T> = RecoilValueReadOnly<T> | RecoilState<T>;
+
+export function isRecoilValue(val: unknown): val is RecoilValue<any>;
+
+/** Utilities */
+
+// bigint not supported yet
+export type Primitive = undefined | null | boolean | number | symbol | string;
+
+export type SerializableParam = Primitive | SerializableParam[] | { [key: string]: SerializableParam };
+
+export interface AtomFamilyOptions<T, P extends SerializableParam> {
+  key: NodeKey;
+  dangerouslyAllowMutability?: boolean;
+  default: RecoilValue<T> | Promise<T> | T | ((param: P) => T | RecoilValue<T> | Promise<T>);
```

Comment:
good point! But I should've mentioned we don't want to introduce any new terminology in this file so that the source of truth lives in flow and this just mirrors that implementation as much as possible (I'll add a comment at the top of the file to ensure that's the case)